### ISSUE TITLE
React Native core wrapper

### ISSUE THREAD

**Issue by ianthetechie** [stadiamaps/ferrostar#116]
Wrapping the `FerrostarCore` classes in an Expo library will get us halfway to full RN support. MapLibre React Native already provides an acceptable solution for a fully custom UI.

There is an ongoing [slack thread](https://osmus.slack.com/archives/C065DB4T2UB/p1717770856348039?thread_ts=1716425877.458769&cid=C065DB4T2UB) discussing the process, and it appears that RN doesn't support SPM: https://github.com/react-native-community/discussions-and-proposals/issues/587.

It appears that a workaround that enables integrating (at least some types of) SPM packages _via CocoaPods_: https://github.com/facebook/react-native/pull/44627. The PR is pretty fresh and the last comment was only a few days ago, so I'm somewhat hopeful this can land (not sure how long it takes to get to a release though).

**Comment by ianthetechie** [comment]
Dropping this link as well in case it becomes relevant: https://stackoverflow.com/questions/77701620/how-to-support-swift-macro-in-a-project-integrated-by-cocoapods/. This _shouldn't_ be an issue for the core, but it might be depending on how the pod works.

**Comment by ianthetechie** [comment]
The PR has been merged, but is not yet in a release. It looked like it would land in 0.75, but it appears to have not progressed from the release candidates to the final release. We'll keep an eye on the 0.76 release.

**Comment by ianthetechie** [comment]
Today I became aware of a cool project that's working toward solving this: https://github.com/jhugman/uniffi-bindgen-react-native!

**Comment by bjtrounson** [comment]
Leaving this here for anyone who might be looking for a React Native support. 

I have a expo modules wrapper in the works locate here: https://github.com/bjtrounson/expo-ferrostar

This wrapper only currently supports Android but I'm looking at iOS support

As mentioned above react-native doesn't support SPM natively at the moment. This PR has been merge (https://github.com/facebook/react-native/pull/44627) which gives SPM support via the podspec file.

I've seen other issues of it being used so I will have a crack at using RN 0.76.0

**Comment by bjtrounson** [comment]
> Did you have to generate all those [record models](https://github.com/bjtrounson/expo-ferrostar/tree/main/android/src/main/java/expo/modules/ferrostar/records) yourself 🤯? Or are there some auto-generation tools?

Yes I created them manually, I don't think there is a auto-generation tool for expo module records (would really like to have one lol.)

> I hadn't even considered whether WASM was a possible approach to bridging the core 🤔 We already support WASM as a build target and have a working web components frontend. Maybe that could simplify bridging the core (it even has TypeScript Bindings) into React Native, and bypass all of the issues that come with native packaging?

I don't know much about WASM binding or Rust so can't really comment lol. I saw this project mention there, this looks like a nice template to create Turbo Modules via that rust binding [https://github.com/unomed-dev/react-native-matrix-sdk](https://github.com/unomed-dev/react-native-matrix-sdk). I don't know how you would interact with WASM via the Turbo Modules though as this seems to be only interacting to Rust.
